### PR TITLE
[risk=no] Improve cleanBuffer CREATING -> ERROR transition logic

### PR DESCRIPTION
**Pull request by calbach**
Per slack discussion, comparing to the current time may not be correct given:
- large sets of billing projects (syncBillingProjectBuffer may not visit all projects within the deadline)
- Firecloud API errors, or extended outages

Transitions in either of these cases would result in leaked billing projects.

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 1]
```diff
@@ -137,13 +140,34 @@ public void syncBillingProjectStatus() {
...
+                    false)
+                // Ensure that we've also synchronized the status since the deadline elapsed. This
+                // covers degenerate cases where syncBillingProjectStatus might be backlogged and
+                // hasn't caught up to check all the CREATING projects, or covers periods where
+                // the Firecloud API is unresponsive. Ideally we'd push this logic directly down
+                // into the DAO, but JPQL has poor support for date comparisons. Keep the above
+                // filter as well as a first pass filter to limit what we pull into memory.
+                .filter(
+                    entry ->
+                        entry.getLastSyncRequestTime() != null
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: nit: I'd do `Optional.ofNullable(entry.getLastSyncRequestTime()).map(...).flatMap(Streams::stream)` (to squash out the empty optionals). I suppose this is primarily a style preference though, but I feel like null checks are a bit of a smell in stream logic.

Better yet, I'd likely just make a helper function like `isBufferEntryInTimespan()` and filter on that using a method reference. This would also make the filter function(s) separately testable.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 2]
```diff
@@ -137,13 +140,34 @@ public void syncBillingProjectStatus() {
 
   public void cleanBillingBuffer() {
     Iterables.concat(
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: nit: I'm finding this hard to follow, since the arguments to concat() are so involved and far apart. Why not declare two lists separately and concat them at the end? Seems like that's a win for debuggability too.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 3]
```diff
@@ -479,13 +479,26 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(entry.getFireCloudProjectName());
+    billingProjectBufferService.syncBillingProjectStatus();
+
     CLOCK.setInstant(NOW.plus(61, ChronoUnit.MINUTES));
     billingProjectBufferService.cleanBillingBuffer();
 
+    // Time has elapsed but no status sync has occurred, should not transition.
+    assertThat(billingProjectBufferEntryDao.findOne(entry.getId()).getStatusEnum())
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: I'm not sure how much of your logic is really getting exercised here. Do you only have a single entry with CREATING status? I think you'd want cases where (1) they're all before the time range starts, (2) some are on each side (3) one is exactly on the border, etc. That is, simulate a stream of them and operate on that.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 4]
```diff
@@ -479,13 +479,26 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
+        .isEqualTo(CREATING);
+
+    // Sync again, now @ +61 minutes - should transition.
+    billingProjectBufferService.syncBillingProjectStatus();
+    billingProjectBufferService.cleanBillingBuffer();
     assertThat(billingProjectBufferEntryDao.findOne(entry.getId()).getStatusEnum())
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Are you asserting that the first one you find is an error? Should they not all be in the error state?

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 3 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Thanks, I think the boundary handling is the easiest place to write a bug, so that's good to test.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java (hunk 2 above)
Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: So I think part of the issue is that I trained myself to try to read stream statements in one mental "bite" but I can't really do that here. Which may be fine. In other words, if most code is stream operations, this will become more natural.

The biggest issue reading it for me was that I couldn't see the whole statement at once on my laptop, so I have to remember context (like the iterables.concat() around everything), try to find the comma, etc.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 5]
```diff
@@ -136,21 +141,15 @@ public void syncBillingProjectStatus() {
...
-                DbStorageEnums.billingProjectBufferStatusToStorage(CREATING),
-                new Timestamp(
-                    clock
-                        .instant()
-                        .minus(CREATING_TIMEOUT_MINUTES, ChronoUnit.MINUTES)
-                        .toEpochMilli())),
+            findExpiredCreatingEntries(now),
+            // For the ASSIGNING status monitor, we can simply filter by the current time as this
+            // status tracks internal state rather than the Firecloud status.
             billingProjectBufferEntryDao.findAllByStatusAndLastStatusChangedTimeLessThan(
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: nit: Aesthetically I might change "TimeLessThan" to "EarlierThan" in this name. Just because I don't think of wall clock time as something comparable in that way.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 6]
```diff
@@ -163,6 +162,32 @@ public void cleanBillingBuffer() {
...
   }
 
+  private List<DbBillingProjectBufferEntry> findExpiredCreatingEntries(Instant now) {
+    return StreamSupport.stream(
+            billingProjectBufferEntryDao
+                .findAllByStatusAndLastStatusChangedTimeLessThan(
+                    DbStorageEnums.billingProjectBufferStatusToStorage(CREATING),
+                    new Timestamp(
+                        now.minus(CREATING_TIMEOUT_MINUTES, ChronoUnit.MINUTES).toEpochMilli()))
+                .spliterator(),
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: IntelliJ tells me you can just do 
```
  private List<DbBillingProjectBufferEntry> findExpiredCreatingEntries(Instant now) {
    return billingProjectBufferEntryDao
        .findAllByStatusAndLastStatusChangedTimeLessThan(
            DbStorageEnums.billingProjectBufferStatusToStorage(CREATING),
            new Timestamp(
                now.minus(CREATING_TIMEOUT_MINUTES, ChronoUnit.MINUTES).toEpochMilli()))
        .stream()
// ...
```

Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: also nit: I'd probably declare the cutoff timestamp separately, since it's a  slightly involved construction and calculation.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 7]
```diff
@@ -479,13 +481,26 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
     entry.setStatusEnum(CREATING, this::getCurrentTimestamp);
     entry.setCreationTime(getCurrentTimestamp());
     billingProjectBufferEntryDao.save(entry);
 
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(entry.getFireCloudProjectName());
+    billingProjectBufferService.syncBillingProjectStatus();
+
     CLOCK.setInstant(NOW.plus(61, ChronoUnit.MINUTES));
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: I'd define the number of minutes in terms of the  grace period instead of just saying 61. It would be easy to change the setting and have the test get out of date (though in most scenarios hopefully the test would start breaking).

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 8]
```diff
@@ -479,13 +481,26 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
     billingProjectBufferService.cleanBillingBuffer();
 
+    // Time has elapsed but no status sync has occurred, should not transition.
+    assertThat(billingProjectBufferEntryDao.findOne(entry.getId()).getStatusEnum())
+        .isEqualTo(CREATING);
+
+    // Sync again, now @ +61 minutes - should transition.
+    billingProjectBufferService.syncBillingProjectStatus();
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Should we always call sync before clean? If so, why not have the latter call the former for us? 

I'm starting to think this service may be exhibiting Temporal Coupling. It's hard to understand what the invariants are w/r/t the billing buffer.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 9]
```diff
@@ -163,6 +162,32 @@ public void cleanBillingBuffer() {
...
+        // into the DAO, but JPQL has poor support for date comparisons. Keep the above
+        // filter as well as a first pass filter to limit what we pull into memory.
+        .filter(entry -> entry.getLastSyncRequestTime() != null)
+        .filter(
+            entry ->
+                Duration.between(
+                            entry.getLastStatusChangedTime().toInstant(),
+                            entry.getLastSyncRequestTime().toInstant())
+                        .toMinutes()
+                    >= CREATING_TIMEOUT_MINUTES)
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: Yeah, if your specified timeout is itself a Duration, then you don't have to get into units here. I don't  know if the config system supports durations yet though.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 10]
```diff
@@ -163,6 +162,32 @@ public void cleanBillingBuffer() {
...
+                        now.minus(CREATING_TIMEOUT_MINUTES, ChronoUnit.MINUTES).toEpochMilli()))
+                .spliterator(),
+            false)
+        // Ensure that we've also synchronized the status since the deadline elapsed. This
+        // covers degenerate cases where syncBillingProjectStatus might be backlogged and
+        // hasn't caught up to check all the CREATING projects, or covers periods where
+        // the Firecloud API is unresponsive. Ideally we'd push this logic directly down
+        // into the DAO, but JPQL has poor support for date comparisons. Keep the above
+        // filter as well as a first pass filter to limit what we pull into memory.
+        .filter(entry -> entry.getLastSyncRequestTime() != null)
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: do we not also need to check `getLastStatusChangedTime()` for nullity?

Also, could you put a predicate method on `DbBillingProjectBufferEntry` that does this computation  as a derived property? Then it's just `.filter(DbBillingProjectBufferEntry::isInExpiredCreatingState)`. (May want to pass in the timestamp for testing though.)

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 11]
```diff
@@ -479,13 +481,26 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
   @Test
   public void cleanBillingBuffer_creating() {
     DbBillingProjectBufferEntry entry = new DbBillingProjectBufferEntry();
+    entry.setFireCloudProjectName("foo");
     entry.setStatusEnum(CREATING, this::getCurrentTimestamp);
     entry.setCreationTime(getCurrentTimestamp());
     billingProjectBufferEntryDao.save(entry);
 
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: +1 for `doReturn()`. It's especially important if you need to return `CompletableFuture`s.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 3 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: I don't see where this entry was ever in a state other than CREATING.

If the point is that it stays in creating, then it would be good to see a test with a different timeout such that the opposite is true. I'm still getting context on the overall mission here, so I may be confused.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 12]
```diff
@@ -529,6 +544,55 @@ public void cleanBillingBuffer_ignoreValidEntries() {
...
+    billingProjectBufferService.bufferBillingProjects();
+
+    CLOCK.setInstant(NOW.plus(30, ChronoUnit.MINUTES));
+    billingProjectBufferService.bufferBillingProjects();
+    billingProjectBufferService.bufferBillingProjects();
+
+    final ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+    verify(fireCloudService, times(3)).createAllOfUsBillingProject(captor.capture());
+    final List<String> capturedProjectNames = captor.getAllValues();
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: The need to have a mocked return paired with each input may indicate that we'd benefit from a stand-along test fixture that could take in a map (for example). If there were such a fixture for the billing buffer, I might have a small use for it in the billing status PR test.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 13]
```diff
@@ -529,6 +544,55 @@ public void cleanBillingBuffer_ignoreValidEntries() {
...
+        .when(fireCloudService)
+        .getBillingProjectStatus(capturedProjectNames.get(0));
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(capturedProjectNames.get(1));
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.ERROR))
+        .when(fireCloudService)
+        .getBillingProjectStatus(capturedProjectNames.get(2));
+
+    Supplier<List<BillingProjectBufferStatus>> statuses =
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: I like it.

I'd probably name  this something like `capturedProjectsStatusSupplier` so it sticks out.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 8 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: sync processes a limited number of entries on every run. Yes, it was exhibiting Temporal Coupling, though probably not intentionally. That is the bug this PR is fixing. I'm not aware of any more instances of that pattern in the current buffer system.

Here is the original design doc: http://broad.io/1ppw, it doesn't contain any details about the different cronjobs. Brief summary which we should probably writeup as a documentation task:

Crons:
bufferBillingProjects (every minute): create new entries in the buffer, always in the CREATING state
syncBillingProjectStatus (every minute): poll for state changes on the 5 least recently synced entries in a relevant state, namely tracking transitions from CREATING --> ACTIVE or ERROR
cleanBillingBuffer (every hour): clean bad entries out of the buffer, e.g. things that have been in a particular state for too long; this is necessary as the buffer has a fixed capacity

Methods (buffer consumer):
create/clone workspace: mark an entry as ASSIGNING, then ASSIGNED when used for a workspace. If we fail in the middle, cleanBillingBuffer cleans out entries stuck in ASSIGNING

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 14]
```diff
@@ -136,21 +141,15 @@ public void syncBillingProjectStatus() {
...
+            // For the ASSIGNING status monitor, we can simply filter by the current time as this
+            // status tracks internal state rather than the Firecloud status.
             billingProjectBufferEntryDao.findAllByStatusAndLastStatusChangedTimeLessThan(
                 DbStorageEnums.billingProjectBufferStatusToStorage(ASSIGNING),
                 new Timestamp(
-                    clock
-                        .instant()
-                        .minus(ASSIGNING_TIMEOUT_MINUTES, ChronoUnit.MINUTES)
-                        .toEpochMilli())))
+                    now.minus(ASSIGNING_TIMEOUT_MINUTES, ChronoUnit.MINUTES).toEpochMilli())))
```
[the remark this answers is not in the document]
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: Yes - good suggestion.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 3 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: It starts in CREATING and then it moves to ERROR when the time limit expires. That second status check happens later at L505. I split this into two tests to cover the normal path, and the delayed sync path separately

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Personal preference, but I don't think it's necessary to go full Java verbosity mode for a tightly scoped local variable. Will at least call it `statusSupplier` to make it a little clearer though.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 12 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: An in-memory fake for the Firecloud API would probably be useful.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 7 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Always need to balance this against test complexity, but in this situation I think it's an improvement - done

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 15]
```diff
@@ -479,13 +489,45 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
+    entry.setCreationTime(getCurrentTimestamp());
+    billingProjectBufferEntryDao.save(entry);
+
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(entry.getFireCloudProjectName());
+    billingProjectBufferService.syncBillingProjectStatus();
+
+    CLOCK.setInstant(AFTER_CREATING_TIMEOUT);
+    billingProjectBufferService.cleanBillingBuffer();
```
Comment (ericsong) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: do we have a test case that verifies entries are not removed if `sync` has never been called? It feels like a slightly different test case.

Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: True, worth covering - added.

**Comment by calbach** [comment]
FYI @jaycarlton I'll wait for reapproval from you here since there were a fair number of changes/replies

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java (hunk 10 above)
Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: OK, I guess lots of folks prefer smart services and dumb entities.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 16]
```diff
@@ -479,13 +489,65 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
+    entry.setFireCloudProjectName("foo");
+    entry.setStatusEnum(CREATING, this::getCurrentTimestamp);
+    entry.setCreationTime(getCurrentTimestamp());
+    billingProjectBufferEntryDao.save(entry);
+
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(entry.getFireCloudProjectName());
+
+    CLOCK.setInstant(AFTER_CREATING_TIMEOUT);
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: nit: "after creating timeout" can be parsed two ways. I initially read it as after (creating timeout), meaning after it was started, rather than after it was expired. Maybe `AFTER_CREATING_TIMEOUT_ELAPSED` would be clearer.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 8 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Awesome, thanks. That's helpful. One thing that jumps out is, why don't we consolidate sync with clean, since both are concerned with updating status based on elapsed time.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 12 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Let's start making a list of such things... It'd be good to settle on some fake design principles potentially.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 17]
```diff
@@ -506,27 +568,76 @@ public void cleanBillingBuffer_assigning() {
...
+        .when(fireCloudService)
+        .getBillingProjectStatus(capturedProjectNames.get(0));
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(capturedProjectNames.get(1));
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.ERROR))
+        .when(fireCloudService)
+        .getBillingProjectStatus(capturedProjectNames.get(2));
+
+    Supplier<List<BillingProjectBufferStatus>> statusSupplier =
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Since this is a fairly new pattern, a little comment here would be helpful I think.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 4 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: That's probably good enough (since you are familiar with the implementation), but I'd want to make sure it's operating on all and only the entries it's supposed to. It's easy to imagine a bug that happened only when more than 1 (or 5) entries were present.

Not a task for this PR though.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 8 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: In theory you could combine all three of these cron jobs. They all have different roles, and I think it is a better design to allow these pieces to operate independently both for failure isolation, Firecloud load, and to allow entries to move quickly through the state machine. cleanBillingBuffer is also not scoped exclusively to Firecloud project status changes

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 4 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: That's the other test case that I added in response to your other comment: `cleanBillingBuffer_multipleCreates`. I'm not sure what more you're looking for in this test case.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 16 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Done

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 17 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Done

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 4 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: Oh, I'm sorry. I lost track of what was new.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 18]
```diff
@@ -33,8 +37,8 @@
 
   private static final int SYNCS_PER_INVOCATION = 5;
   private static final int PROJECT_BILLING_ID_SIZE = 8;
-  private static final int CREATING_TIMEOUT_MINUTES = 60;
-  private static final int ASSIGNING_TIMEOUT_MINUTES = 10;
+  @VisibleForTesting static final Duration CREATING_TIMEOUT = Duration.ofMinutes(60);
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: 👏

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 3 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: It occurs to me that if we need to add anything else to that enum, we should probably break out the OK/ERROR status into a separate one, and last known good status in the known transition of states. As it stands, we don't have a way to indicate more than one kind of error (e.g. FAILED_TO_GARBAGE_COLLECT). Right now everything in ERROR is assumed to be a creation failure from what I can tell. I also tend to add a catch-all UNKNOWN status for things being migrated or batch processed to indicated unexpected exceptions, interruptions, etc.

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java [hunk 19]
```diff
@@ -479,13 +489,65 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
...
+    entry.setCreationTime(getCurrentTimestamp());
+    billingProjectBufferEntryDao.save(entry);
+
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(entry.getFireCloudProjectName());
+
+    CLOCK.setInstant(AFTER_CREATING_TIMEOUT_ELAPSED);
+    billingProjectBufferService.syncBillingProjectStatus();
+    billingProjectBufferService.cleanBillingBuffer();
```
Comment (jaycarlton) [inline api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: If we expect, in reality, that some more time elapses between the sync and the clean, then it wouldn't hurt to advance the clock some in between these two events. It could potentially catch a bug if somehow things were better when the two events were nearly simultaneous. (Though I imagine it's more common for race conditions to happen when events like this are happening near each other.)

Bonus question: is it safe for more than one of these crons to be operating on an entry at the same time? I haven't looked at how they're scheduled to see if that's possible.

File: api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java [hunk 20]
```diff
@@ -160,6 +156,30 @@ public void cleanBillingBuffer() {
...
+        // into the DAO, but JPQL has poor support for date comparisons. Keep the above
+        // filter as well as a first pass filter to limit what we pull into memory.
+        .filter(entry -> entry.getLastStatusChangedTime() != null)
+        .filter(entry -> entry.getLastSyncRequestTime() != null)
+        .filter(
+            entry ->
+                Duration.between(
+                            entry.getLastStatusChangedTime().toInstant(),
+                            entry.getLastSyncRequestTime().toInstant())
+                        .toMillis()
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: nit: if you prefer, [there's also `Duration.compareTo()`](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#compareTo-java.time.Duration-), which I think can be woven directly into the filter somehow without converting both to millis.

Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java]: I had this initially, but decided it was unreadable; directly comparing these values with a >= sign seems much better to me (would be nice if there were another Duration method or util to do this directly).

File: api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java (hunk 19 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java]: I'd rather keep these tests lean and on-topic. These crons are intended to be safe to operate in parallel (and yes, it is possible). In my view this is sufficient to cover the regression as written, though clearly you could go much further here and add coverage of all permutations of clean/sync ordering with varying time between each.
